orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -244,62 +291,384 @@ Status AddOrSkipRecordForwardWaitBackward(Graph& graph,
-  std::string& backward_recorded_output_name) {
-  // insert WaitEvent and RecordEvent to the partition
-  Node* send_fw{nullptr};
-  Node* send_bw{nullptr};
-  Node* recv_fw{nullptr};
-  Node* recv_bw{nullptr};
+  std::string& forward_wait_output_name,
+  std::string& forward_record_output_name,
+  std::string& backward_wait_output_name,
+  std::string& backward_record_output_name,
+  std::string& forward_waited_event_after_recv_name,
+  std::string& forward_recorded_event_before_send_name,
+  std::string& backward_waited_event_after_recv_name,
+  std::string& backward_recorded_event_before_send_name) {
+  // Declare nodes according to their topological order.
+  Node* forward_wait{nullptr};
+  Node* forward_send{nullptr};
+  Node* forward_recv{nullptr};
+  Node* backward_recv{nullptr};
+  Node* backward_send{nullptr};
+  Node* backward_record{nullptr};
   for (auto& node : graph.Nodes()) {
     if (node.OpType() == "Send") {
       if (IsBackward(node)) {
-        send_bw = &node;
+        backward_send = &node;
       } else {
-        send_fw = &node;
+        forward_send = &node;
       }
     } else if (node.OpType() == "Recv") {
       if (IsBackward(node)) {
-        recv_bw = &node;
+        backward_recv = &node;
       } else {
-        recv_fw = &node;
+        forward_recv = &node;
       }
     }
   }
 
+  // Names to added into this graph's input list.
+  // Their value may be provides as "feeds" when calling session.Run(...).
   std::vector<std::string> new_input_names;
+  // Names to added into this graph's output list.
+  // Their value may be provides as "feeds" when calling session.Run(...).
   std::vector<std::string> new_output_names;
 
-  ORT_RETURN_IF_ERROR(AddRecordBackward(
+  backward_record = AddRecordBackward(
     graph,
-    send_bw,
+    backward_send,
     new_input_names,
     new_output_names,
     backward_recorded_event_name,
-    backward_recorded_output_name));
-  ORT_RETURN_IF_ERROR(AddWaitForward(
+    backward_record_output_name);
+  forward_wait = AddWaitForward(
     graph,
-    recv_fw,
+    forward_recv,
     new_input_names,
     forward_waited_event_name,
-    forward_waited_output_name));
+    forward_wait_output_name);
   ORT_RETURN_IF_ERROR(AddOrSkipRecordForwardWaitBackward(
     graph,
-    send_fw,
-    recv_bw,
+    forward_send,
+    backward_recv,
     new_input_names,
     forward_recorded_event_name,
     backward_waited_event_name,
-    forward_recorded_output_name,
-    backward_waited_output_name));
+    forward_record_output_name,
+    backward_wait_output_name));
+  
+  // Different stages have different patterns of Send & Recv.
+  // For different patterns, we add different WaitEvent and Record.
+  //
+  // After applying all transformations above, the current patterns
+  // are listed below.
+  //
+  // 1. First stage:
+  //  Wait -----------------> FW -> --------> Send -> Record ->
+  //  Wait -> Recv ---------> BW -> ----------------> Record
```

Comment:
Might be easier for reader to understand, if we differentiate WaitData vs. WaitCompute/WaitPipeline

Reply:
Ok. I will try to find a place to put an overview message in my next PR. If we put them here, those comments may be prolonged significantly.